README.md
battery-icon
============

Simple battery monitor inspired by volumeicon.

Why not name it batteryicon? Because there's software called like that already
(some funny application for Blackberry).

Why another battery systray plugin? Because I couldn't find anything that was
simple, functional and wasn't part of a window manager or desktop environment.

Running
=======

Simply './autogen.sh', './configure', 'make' and './battery-icon'.

UPower with its DBus interface must be running on the system.

TODO
====
* ~~autotolize~~ DONE

* ~~add conditional compilation for libnotify~~ DONE

* ~~get rid of g_spawn_* and implement acpi polling natively~~

* implement detailed mode, which would use upower to provide useful battery information

* implement simple preferences

* implement debug mode

* implement support for multiple batteries

* implement Qt version


NEWS
battery-icon 0.2.1
==================

Fix a bug when warning notifications were shown when a battery was charging.

Configure script no longer checks for 'acpi' binary, since it's no longer
needed.


battery-icon 0.2.0
==================

Basic support for one system battery is implemented.

Battery status is monitored using UPower and its DBus interface.


upower.c
/*
 * Copyright (C) 2014 Jan Synáček
 *
 * Author: Jan Synáček <jan.synacek@gmail.com>
 * URL: https://github.com/jsynacek/battery-icon
 * Created: May 2014
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.  If not, write to
 * the Free Software Foundation, Inc., 51 Franklin Street, Fifth
 * Floor, Boston, MA 02110-1301, USA.
 */

#include "upower.h"

GDBusProxy *upower_init_proxy(GCallback callback)
{
	GDBusProxy *proxy = NULL;
	GError *error = NULL;

	proxy = g_dbus_proxy_new_for_bus_sync(G_BUS_TYPE_SYSTEM,
					      G_DBUS_PROXY_FLAGS_NONE,
					      NULL,
					      "org.freedesktop.UPower",
					      "/org/freedesktop/UPower/devices/battery_BAT0",
					      "org.freedesktop.UPower.Device",
					      NULL, &error);
	g_signal_connect(G_OBJECT(proxy), "g-signal",
			 callback, NULL);

	return proxy;
}

void upower_destroy_proxy(GDBusProxy *proxy)
{
	if (proxy) {
		g_object_unref(proxy);
	}
}

GVariant *upower_get_property(GDBusProxy *proxy, const gchar *property)
{
	GVariant *result;
	GVariant *value = NULL;
	GError   *error = NULL;

	result = g_dbus_proxy_call_sync(proxy,
					"org.freedesktop.DBus.Properties.Get",
					g_variant_new("(ss)", "org.freedesktop.UPower.Device", property),
					G_DBUS_CALL_FLAGS_NONE, -1, NULL,
					&error);
	if (error) {
		g_printerr(error->message);
		return NULL;
	}

	g_variant_get(result, "(v)", &value);
	return value;
}

gdouble upower_get_percentage(GDBusProxy *proxy)
{
	GVariant *value = upower_get_property(proxy, "Percentage");
	gdouble result = g_variant_get_double(value);
	g_variant_unref(value);
	return result;
}

gint64 upower_get_time_to_empty(GDBusProxy *proxy)
{
	GVariant *value = upower_get_property(proxy, "TimeToEmpty");
	guint64 result = g_variant_get_int64(value);
	g_variant_unref(value);
	return result;
}

gint64 upower_get_time_to_full(GDBusProxy *proxy)
{
	GVariant *value = upower_get_property(proxy, "TimeToFull");
	gint64 result = g_variant_get_int64(value);
	g_variant_unref(value);
	return result;
}

gint64 upower_get_state(GDBusProxy *proxy)
{
	GVariant *value = upower_get_property(proxy, "State");
	guint32 result = g_variant_get_uint32(value);
	g_variant_unref(value);
	return result;
}

/* TODO multiple batteries support

 $ dbus-send --print-reply \
             --system \
             --dest=org.freedesktop.UPower \
             /org/freedesktop/UPower \
             org.freedesktop.UPower.EnumerateDevices


 GError *error = NULL;
 GVariant *result = NULL;
 GDBusProxy *proxy = g_dbus_proxy_new_for_bus_sync(G_BUS_TYPE_SYSTEM,
 				      G_DBUS_PROXY_FLAGS_NONE,
 				      NULL,
 				      "org.freedesktop.UPower",
 				      "/org/freedesktop/UPower",
 				      "org.freedesktop.UPower",
 				      NULL, &error);
 if (!proxy) {
 	g_printerr("could'nt get proxy");
 	return 1;
 }
 error = NULL;
 result = g_dbus_proxy_call_sync(proxy,
 				"EnumerateDevices", NULL,
 				G_DBUS_CALL_FLAGS_NONE, -1, NULL,
 				&error);
 if (error) {
 	g_printerr(error->message);
 	return 1;
 }
 GVariantIter *iter = NULL;
 g_variant_get(result, "(ao)", &iter);
 gchar *str;
 while (g_variant_iter_loop(iter, "o", &str)) {
 	g_print("battery: %s\n", str);
 }

 g_variant_iter_free(iter);
*/


upower.h
/*
 * Copyright (C) 2014 Jan Synáček
 *
 * Author: Jan Synáček <jan.synacek@gmail.com>
 * URL: https://github.com/jsynacek/battery-icon
 * Created: May 2014
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.  If not, write to
 * the Free Software Foundation, Inc., 51 Franklin Street, Fifth
 * Floor, Boston, MA 02110-1301, USA.
 */

#ifndef _UPOWER_H_
#define _UPOWER_H_

#include <gtk/gtk.h>

GDBusProxy *upower_init_proxy(GCallback callback);
void upower_destroy_proxy(GDBusProxy *proxy);

GVariant *upower_get_property(GDBusProxy *proxy, const gchar *property);
gdouble upower_get_percentage(GDBusProxy *proxy);
gint64 upower_get_time_to_empty(GDBusProxy *proxy);
gint64 upower_get_time_to_full(GDBusProxy *proxy);
gint64 upower_get_state(GDBusProxy *proxy);

#endif


battery.h
/*
 * Copyright (C) 2014 Jan Synáček
 *
 * Author: Jan Synáček <jan.synacek@gmail.com>
 * URL: https://github.com/jsynacek/battery-icon
 * Created: May 2014
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.  If not, write to
 * the Free Software Foundation, Inc., 51 Franklin Street, Fifth
 * Floor, Boston, MA 02110-1301, USA.
 */

#ifndef _BATTERY_H_
#define _BATTERY_H_

#include <gtk/gtk.h>

enum {
	BATTERY_EMPTY,
	BATTERY_CRITICAL,
	BATTERY_LOW,
	BATTERY_MEDIUM,
	BATTERY_HIGH,
	BATTERY_FULL,
	BATTERY_CHARGING,
	BATTERY_CHARGED = 12
};

/* battery states corresponding to state values from upower device */
enum {
	BATTERY_STATE_UNKNOWN,
	BATTERY_STATE_CHARGING,
	BATTERY_STATE_DISCHARGING,
	BATTERY_STATE_EMPTY,
	BATTERY_STATE_FULLY_CHARGED,
	BATTERY_STATE_PENDING_CHARGE,
	BATTERY_STATE_PENDING_DISCHARGE
};

typedef struct battery_info {
	/* dbus properties */
	gdouble percentage;
	guint32 state;
	gint64  time_to_full;
	gint64  time_to_empty;
	/* gui */
	const gchar *notification_status;
	const gchar *icon_name;
	gchar *tooltip;
} battery_info_t;

extern const gchar *battery_icons_primary[];
extern const gchar *battery_icons_default[];
extern const gchar *battery_status_messages[];

battery_info_t *battery_info_init_from_dbus(GDBusProxy *proxy);
void battery_info_destroy(battery_info_t *binfo);

#endif


battery-icon.c
/*
 * Copyright (C) 2014 Jan Synáček
 *
 * Author: Jan Synáček <jan.synacek@gmail.com>
 * URL: https://github.com/jsynacek/battery-icon
 * Created: May 2014
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.  If not, write to
 * the Free Software Foundation, Inc., 51 Franklin Street, Fifth
 * Floor, Boston, MA 02110-1301, USA.
 */

#include <stdlib.h>
#include <gtk/gtk.h>
#include <libnotify/notify.h>
#include "battery.h"
#include "upower.h"
#include "config.h"

static GtkStatusIcon *status_icon = NULL;
#ifdef COMPILEWITH_NOTIFY
static NotifyNotification *notification = NULL;
#endif


static void status_icon_on_popup_menu(GtkStatusIcon *status_icon, guint button,
				      guint activation_time, gpointer user_data)
{
	GtkWidget *gtkMenu = gtk_menu_new();
	GtkWidget *quit = gtk_image_menu_item_new_from_stock("gtk-quit", NULL);

	g_signal_connect(G_OBJECT(quit), "activate",
			 G_CALLBACK(gtk_main_quit), NULL);

	gtk_menu_shell_append(GTK_MENU_SHELL(gtkMenu), quit);

	gtk_widget_show_all(gtkMenu);

	gtk_menu_popup(GTK_MENU(gtkMenu), NULL, NULL,
		       gtk_status_icon_position_menu, status_icon,
		       button, activation_time);
}
void on_upower_changed(GDBusProxy *proxy,
		       gchar      *sender_name,
		       gchar      *signal_name,
		       GVariant   *parameters,
		       gpointer    user_data)
{
	battery_info_t *binfo = battery_info_init_from_dbus(proxy);

 	gtk_status_icon_set_tooltip_text(status_icon, binfo->tooltip);
	gtk_status_icon_set_from_icon_name(status_icon, binfo->icon_name);

#ifdef COMPILEWITH_NOTIFY
	if (binfo->percentage < 20 && binfo->state != BATTERY_STATE_CHARGING) {
		GError *error = NULL;
		notify_notification_update(notification,
					   binfo->notification_status,
					   binfo->tooltip,
					   binfo->icon_name);
		notify_notification_show(notification, &error);
		if(error) {
			g_printerr("%s\n", error->message);
		}
	}
#endif

	/* DEBUG */
	g_print("tooltip: \"%s\"\n", binfo->tooltip);

	battery_info_destroy(binfo);
}
int main(int argc, char *argv[])
{
	gtk_init(&argc, &argv);
	status_icon = gtk_status_icon_new_from_icon_name("gpm-battery-charged");
	gtk_status_icon_set_visible(status_icon, TRUE);
	g_signal_connect(G_OBJECT(status_icon), "popup-menu",
			 G_CALLBACK(status_icon_on_popup_menu), NULL);

	GDBusProxy *proxy = upower_init_proxy(G_CALLBACK(on_upower_changed));

#ifdef COMPILEWITH_NOTIFY
	notify_init(PACKAGE_STRING);
	notification = notify_notification_new(NULL, NULL, NULL);
#endif

	/* initialize for the first time */
	battery_info_t *binfo = battery_info_init_from_dbus(proxy);
 	gtk_status_icon_set_tooltip_text(status_icon, binfo->tooltip);
	gtk_status_icon_set_from_icon_name(status_icon, binfo->icon_name);
	battery_info_destroy(binfo);

	gtk_main();

	upower_destroy_proxy(proxy);

#ifdef COMPILEWITH_NOTIFY
	notify_uninit();
#endif
	return EXIT_SUCCESS;
}


battery.c
/*
 * Copyright (C) 2014 Jan Synáček
 *
 * Author: Jan Synáček <jan.synacek@gmail.com>
 * URL: https://github.com/jsynacek/battery-icon
 * Created: May 2014
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; see the file COPYING.  If not, write to
 * the Free Software Foundation, Inc., 51 Franklin Street, Fifth
 * Floor, Boston, MA 02110-1301, USA.
 */

#include "battery.h"
#include "upower.h"

const gchar *battery_icons_primary[] = {
	"gpm-primary-000",
	"gpm-primary-020",
	"gpm-primary-040",
	"gpm-primary-060",
	"gpm-primary-080",
	"gpm-primary-100",
	"gpm-primary-000-charging",
	"gpm-primary-020-charging",
	"gpm-primary-040-charging",
	"gpm-primary-060-charging",
	"gpm-primary-080-charging",
	"gpm-primary-100-charging",
	"gpm-primary-charged",
	NULL
};

const gchar *battery_icons_default[] = {
	"battery-empty",
	"battery-caution",
	"battery-low",
	"battery-good",
	/* defaults have only 5 levels, so use "good" twice */
	"battery-good"
	"battery-full",
	"battery-empty-charging",
	"battery-caution-charging",
	"battery-low-charging",
	"battery-good-charging",
	"battery-good-charging",
	"battery-full-charging",
	"battery-full-charged",
	NULL
};

const gchar *battery_status_messages[] = {
	"Battery empty.",
	"Battery critical.",
	"Battery low.",
	"Battery medium.",
	"Battery high.",
	"Battery full.",
	"Battery charging."
};

/* requires percentage and state to be already set */
static void battery_update_gui_items(battery_info_t *binfo)
{
 	GtkIconTheme *icon_theme = gtk_icon_theme_get_default();
 	const gchar **battery_icons;
 	gint info_index;

 	if (gtk_icon_theme_has_icon(icon_theme, battery_icons_primary[0])) {
 		battery_icons = battery_icons_primary;
 	}
 	else {
 		battery_icons = battery_icons_default;
 	}

 	gint percentage = (gint)binfo->percentage; /* TODO possible rounding error */
 	gboolean is_charging = binfo->state == BATTERY_STATE_CHARGING;

 	if (percentage >= 99 && is_charging) {
 		info_index = BATTERY_CHARGED;
 	}
 	else if (percentage >= 90) {
 		info_index = BATTERY_FULL;
 	}
 	else if (percentage >= 70) {
 		info_index = BATTERY_HIGH;
 	}
 	else if (percentage >= 50) {
 		info_index = BATTERY_MEDIUM;
 	}
 	else if (percentage >= 30) {
 		info_index = BATTERY_LOW;
 	}
 	else if (percentage >= 10) {
 		info_index = BATTERY_CRITICAL;
 	}
 	else {
 		info_index = BATTERY_EMPTY;
 	}

 	if (percentage < 99 && is_charging == TRUE) {
 		binfo->icon_name = battery_icons[info_index + BATTERY_CHARGING];
 	} else {
 		binfo->icon_name = battery_icons[info_index];
 	}

 	binfo->notification_status = battery_status_messages[info_index];
}

static void time_to_str(gchar *buf, gint64 time)
{
	gint hrs  = time / 3600;
	gint mins = time / 60 - (hrs * 60);
	sprintf(buf, "%d:%02d", hrs, mins);
}

battery_info_t *battery_info_init_from_dbus(GDBusProxy *proxy)
{
	/* get stuff from upower first */
	battery_info_t *binfo = g_malloc0(sizeof(battery_info_t));
	binfo->percentage = upower_get_percentage(proxy);
  	binfo->state = upower_get_state(proxy);
  	binfo->time_to_full = upower_get_time_to_full(proxy);
	binfo->time_to_empty = upower_get_time_to_empty(proxy);

	/* set up gtk stuff second */
	battery_update_gui_items(binfo);

	/* tooltip third */
	const gchar *what = "";
	gchar tstr[128];
	gchar out[1024];
	if (binfo->state == BATTERY_STATE_CHARGING) {
		what = "until charged";
		time_to_str(tstr, binfo->time_to_full);
	}
	else if (binfo->state == BATTERY_STATE_DISCHARGING) {
		what = "remaining";
		time_to_str(tstr, binfo->time_to_empty);
	}
	sprintf(out, "%.1f%%, %s %s", binfo->percentage, tstr, what);
	binfo->tooltip = g_strdup(out);

	return binfo;
}

void battery_info_destroy(battery_info_t *binfo)
{
	if (binfo->tooltip) {
		g_free(binfo->tooltip);
	}
	g_free(binfo);
}
